pkg/controller/container-runtime-config/helpers.go

```diff
@@ -478,29 +489,124 @@ func validateUserContainerRuntimeConfig(cfg *mcfgv1.ContainerRuntimeConfig) erro
+			}
+			// Always add the payloadRepo to policy allowed so we can resolve for mirrors for it
+			allowed = append(allowed, payloadRepo)
+			continue
 		}
 		// Was not a match to the registry being used by the payload, so add to valid blocked registries
-		blockedRegs = append(blockedRegs, reg)
+		registriesBlocked = append(registriesBlocked, reg)
+		policyBlocked = append(policyBlocked, reg)
+	}
+	if len(blockErr) > 0 {
+		retErr = fmt.Errorf("error adding %q to blocked registries, cannot block the repository being used by the payload", blockErr)
+	}
+	allowed = append(allowed, imgSpec.RegistrySources.AllowedRegistries...)
+	return registriesBlocked, policyBlocked, allowed, retErr
+}
+
+// payloadRepoHasUnblockedMirror returns true if the payload registry has mirror rules configured for it
+func payloadRepoHasUnblockedMirror(payloadRepo reference.Named, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy, imgSpec *apicfgv1.ImageSpec) (bool, error) {
+	// Create a temp registries.conf file with all the registry inputs given
+	tmpFile, err := createTempRegistriesFile(icspRules, imgSpec)
+	if err != nil {
+		return false, err
+	}
+	defer os.Remove(tmpFile.Name())
+
+	// var payloadMirrors []string
+	// Go through the mirror rules configured and check if a source matches the payload registry
+	// Get the possible pull sources for the payload repo using the temp registries.conf file created above
+	r, err := sysregistriesv2.FindRegistry(&types.SystemContext{SystemRegistriesConfPath: tmpFile.Name()}, payloadRepo.Name())
+	if err != nil {
+		return false, err
+	}
+	sources, err := r.PullSourcesFromReference(payloadRepo)
+	if err != nil {
+		return false, err
+	}
+
+	// Go through the list of pull sources and check that at least one of the payload references is not blocked
+	for _, s := range sources {
+		isUsable := true
+		for _, blockReg := range imgSpec.RegistrySources.BlockedRegistries {
+			if runtimeutils.ScopeIsNestedInsideScope(s.Reference.Name(), blockReg) {
+				isUsable = false
+			}
+		}
+		if isUsable {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// createTempRegistriesFile creates a temporary registries config file to be used in determining the valid blocked and allowed registries
+func createTempRegistriesFile(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy, imgSpec *apicfgv1.ImageSpec) (*os.File, error) {
+	tomlConf := sysregistriesv2.V2RegistriesConf{}
+	if err := registries.EditRegistriesConfig(&tomlConf, imgSpec.RegistrySources.InsecureRegistries, imgSpec.RegistrySources.BlockedRegistries, icspRules); err != nil {
+		return nil, err
+	}
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(tomlConf); err != nil {
+		return nil, err
+	}
+	tmpFile, err := os.CreateTemp("", "regtemp")
+	if err != nil {
+		return nil, err
+	}
+	_, err = tmpFile.Write(newData.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	defer tmpFile.Close()
```

Comment:
This should happen just after `os.CreateTemp` succeeds, so that it is called if `Write` fails.

Reply:
fixed